libbeat/beat/instrumentation.go

```diff
@@ -0,0 +1,207 @@
+type CPUProfiling struct {
+	Enabled  bool          `config:"enabled"`
+	Interval time.Duration `config:"interval" validate:"positive"`
+	Duration time.Duration `config:"duration" validate:"positive"`
+}
+
+// IsEnabled indicates whether instrumentation is enabled
+func (c *InstrumentationConfig) IsEnabled() bool {
+	return c != nil && c.Enabled != nil && *c.Enabled
+}
+
+// IsEnabled indicates whether CPU profiling is enabled
+func (p *CPUProfiling) IsEnabled() bool {
+	return p != nil && p.Enabled
+}
+
+// IsEnabled indicates whether heap profiling is enabled
+func (p *HeapProfiling) IsEnabled() bool {
+	return p != nil && p.Enabled
+}
+
+// HeapProfiling holds config information about heap profiling of the APM Server
+type HeapProfiling struct {
+	Enabled  bool          `config:"enabled"`
+	Interval time.Duration `config:"interval" validate:"positive"`
+}
+
+// CreateInstrumentation configures and returns an instrumentation object for tracing
+func CreateInstrumentation(cfg *common.Config, info Info) (*Instrumentation, error) {
+	if !cfg.HasField("instrumentation") {
+		return nil, nil
+	}
+
+	instrumentation, err := cfg.Child("instrumentation", -1)
+	if err != nil {
+		return nil, nil
+	}
+
+	config := InstrumentationConfig{}
+
+	if instrumentation == nil {
+		instrumentation = common.NewConfig()
+	}
+	err = instrumentation.Unpack(&config)
+
+	if err != nil {
+		return nil, fmt.Errorf("could not create tracer, err: %v", err)
+	}
+
+	return initTracer(config, info)
+}
+
+func initTracer(cfg InstrumentationConfig, info Info) (*Instrumentation, error) {
+
+	logger := logp.NewLogger("tracing")
+
+	if !cfg.IsEnabled() {
+		os.Setenv("ELASTIC_APM_ACTIVE", "false")
+		logger.Infof("APM instrumentation is disabled")
+		return nil, nil
+	} else {
+		os.Setenv("ELASTIC_APM_ACTIVE", "true")
+		logger.Infof("APM instrumentation is enabled")
+	}
+
+	if cfg.Profiling.CPU.IsEnabled() {
+		interval := cfg.Profiling.CPU.Interval
+		duration := cfg.Profiling.CPU.Duration
+		logger.Infof("CPU profiling: every %s for %s", interval, duration)
+		os.Setenv("ELASTIC_APM_CPU_PROFILE_INTERVAL", fmt.Sprintf("%dms", int(interval.Seconds()*1000)))
+		os.Setenv("ELASTIC_APM_CPU_PROFILE_DURATION", fmt.Sprintf("%dms", int(duration.Seconds()*1000)))
+	}
+	if cfg.Profiling.Heap.IsEnabled() {
+		interval := cfg.Profiling.Heap.Interval
+		logger.Infof("Heap profiling: every %s", interval)
+		os.Setenv("ELASTIC_APM_HEAP_PROFILE_INTERVAL", fmt.Sprintf("%dms", int(interval.Seconds()*1000)))
+	}
+
+	var tracerTransport transport.Transport
+	var tracerListener net.Listener
+	if cfg.Hosts != nil {
+		// tracing destined for host explicitly set
+		t, err := transport.NewHTTPTransport()
+
+		if err != nil {
+			return nil, err
+		}
+		t.SetServerURL(cfg.Hosts...)
+		if cfg.APIKey != "" {
+			t.SetAPIKey(cfg.APIKey)
+		} else {
+			t.SetSecretToken(cfg.SecretToken)
+		}
+		tracerTransport = t
+		logger.Infof("self pipe directed to %s", cfg.Hosts)
```

Comment:
```suggestion
		logger.Infof("APM tracer directed to %s", cfg.Hosts)
```

Maybe? "self pipe" sounds a bit unusual